### DISCUSSION TITLE
Interface merging and prototype modification of DisplayObject

### DISCUSSION THREAD

**Discussion by alexvgjm** [pixijs/pixijs#9597] [Q&A] [answered]
Hello, here a user of PixiJS with TypeScript,

_first of all, I must thank all the PixiJS developers and contributors, it's a technology I've recently discovered and I find it especially nice. I am infinitely grateful for all those agnostic design decisions and the ease of integration that this entails._

I have a rather rare case that perhaps concerns JavaScript/TypeScript more than pixi itself. Together with one of my students, we have found it feasible to increase the responsibilities of the DisplayObject. Specifically, in a composition-oriented way, we wanted all DisplayObject and child classes to have a Set of "Updatable" along with an update method. The idea of ​​this is simply to keep a logical loop possibility independent of the rendering loop in all the pixi DisplayObject hierarchy.

Taking advantage of TypeScript's interface merging, I came up with something like this:
```ts
declare module 'pixi.js' {
    interface DisplayObject {
        updatables: Set<Updatable>
        update(delta: number): void
    }
}

DisplayObject.prototype.updatables = new Set()
DisplayObject.prototype.update = function (delta: number) {
    this.updatables.forEach(u => u.update(delta))
}
```

It works especially well, **all classes** that inherit from DisplayObject now have an update method and a set of updatables. The IDE recognizes this and everything works as expected except for one problem: **that said Set is obviously shared**. We can't just assign a Set (or any other type of object) to the prototype, it will just make all instances share the same one.

In other systems this kind of problems are solved via overridable initialization methods. That is, if DisplayObject had an init method that I could override (previously calling the original implementation) or some kind of factory method. It could be used to initialize a Set for each new instance.

Is this the case? Is there a way to modify the construction/initialization behavior of DisplayObject and all its derived classes?

_Subclassing DisplayObject is not an option if we can't get the entire hierarchy that extends from the original to extend the new subclass._

**Comment by bigtimebuddy** [comment, marked as answer]
I think you're very close. I would probably do something like this and add a getter/setter for the updateables.

```ts
declare module 'pixi.js' {
    interface DisplayObject {
        updatables: Set<Updatable>
        update(delta: number): void
    }
}

Object.defineProperty(DisplayObject.prototype, 'updateables', {
    get() {
         if (!this._updateables) {
             this._updateables = new Set();
         }
         return this._updateables;
    },
    set(value) {
        this._updateables = value;
    }
});
DisplayObject.prototype._updatables = null;
DisplayObject.prototype.update = function (delta: number) {
    this.updatables.forEach(u => u.update(delta))
}
```

**Reply by alexvgjm** [reply]
In fact this was exactly our first approach and works as expected but...

🤔 maybe it's an unfounded fear, I'm not very familiar with how JavaScript interpreters work when optimizing, but evaluating on each access whether or not something has been initialized seems a bit overkill to me especially for something that is going to be constantly called by each object on each update.

Do you know if this is actually somehow optimized at runtime after the first if?

PD: I've dealt with a similar lazy loading case in the past with a solution that was to replace getters after first run (not in JavaScript), but this broke some static nature by creating functions that bound to the instance with their own closure (similar to creating arrow functions in JavaScript). I wish I didn't have to fall for such hacky solutions.

**Reply by bigtimebuddy** [reply]
Relatively speaking, one if in the getter is not a big deal. This isn't a hack, as defineProperty is standard JS. As long as you initialize the private prototype variable, V8 shouldn't deoptimize DisplayObject. Basically: you're fine, I wouldn't over-optimize too much.
